Add Wasm AOT steps into WasmApp.targets

steveisok opened the pull request:
AOT can now be triggered through the libraries tests as well as the wasm samples.

Passing `/p:RunAOTCompilation=true` is all that is needed for the libraries test run. The property will be false (for now) by default.

Commit:
Restructured Wasm AOT targets to make it work with tests + sample

Commit:
Shift the aotcompile target around so it's visible. Temporary spot until it finds a home

Commit:
Adds AOT steps into WasmApp.targets

AOT can now be triggered through the libraries tests as well as the wasm samples.

Passing `/p:RunAOTCompilation=true` is all that is needed for the libraries test run.
The property will be false (for now) by default.

vargaz commented on src/mono/netcore/sample/wasm/console/Program.cs:
```diff
@@ -2,11 +2,12 @@
 {
-
+    [DynamicDependency(DynamicallyAccessedMemberTypes.All, "System.Runtime.InteropServices.JavaScript.Runtime", "System.Private.Runtime.InteropServices.JavaScript")] 
     public static async Task<int> Main(string[] args)
```
This should probably be handled somewhere else.

vargaz commented on src/mono/netcore/sample/wasm/console/WasmSample.csproj:
```diff
@@ -3,18 +3,16 @@
-    <SelfContained>true</SelfContained>
     <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
-    <RunAnalyzers>false</RunAnalyzers>
     <RunAOTCompilation Condition="'$(RunAOTCompilation)' == ''">false</RunAOTCompilation>
```
This is there to enable people to modify Program.cs during local development.

vargaz commented on src/mono/netcore/sample/wasm/console/WasmSample.csproj:
```diff
@@ -37,79 +35,17 @@
-      <WasmAppDir>$(AppDir)</WasmAppDir>
+      <WasmAppDir>$(AppBundleDir)</WasmAppDir>
+      <WasmAOTDir>$(MSBuildProjectDirectory)\$(PublishDir)\</WasmAOTDir>
       <WasmMainJSPath>$(MonoProjectRoot)\wasm\runtime-test.js</WasmMainJSPath>
```
This should probably be named something like 'WasmObjDir' or 'WasmBuildDir'.

vargaz commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -1,13 +1,51 @@
     <Error Condition="'$(WasmMainAssemblyPath)' == ''" Text="%24(WasmMainAssemblyPath) property needs to be set" />
     <Error Condition="'$(WasmAppDir)' == ''" Text="%24(WasmAppDir) property needs to be set" />
+    <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(WasmAOTDir)' == ''" Text="%24(WasmAOTDir) property needs to be set" />
     <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
```
This should be named something like WasmEnableAOT.

steveisok replied:
`RunAOTCompilation` is a setting that applies to other platforms as well.

radical committed:
MonoAotCompiler: add more logging, and error handling

radical committed:
aotcompiler: fix disableparallelaot to stop if one fails

radical committed:
WasmApp.targets: refactor a bit to allow resolving deps, even in AOT case

radical committed:
small fixes to the sample project files

Commit:
Tweak browser sample to be able to run aot

Commit:
Add feature defaults for the browser sample

Commit:
Merge remote-tracking branch 'upstream/master' into enable-wasm-aot-tests

mdh1418 commented on src/mono/netcore/sample/wasm/browser/WasmSample.csproj:
```diff
@@ -3,33 +3,53 @@
         <WasmAppDir>$(AppDir)</WasmAppDir>
-        <WasmMainAssemblyPath>bin\WasmSample.dll</WasmMainAssemblyPath>
+        <WasmAOTDir>$(MSBuildProjectDirectory)\$(PublishDir)\</WasmAOTDir>
+        <WasmMainAssemblyPath>$(MSBuildProjectDirectory)\bin\WasmSample.dll</WasmMainAssemblyPath>
```
I think this can be removed, or the one on L44

akoeplinger replied:
Yeah this looks fishy.

Commit:
Add RunAnalyzers back in

vargaz commented:
We should merge this and fix it up later since it conflicts with other PRs.

radical committed:
WasmApp.targets: Add new property $(WasmBuildAppDependsOn)

This helps projects to be sure to run their targets *before* any
WasmBuildApp stuff runs, instead of `BeforeTargets` which ends up being
dependent on the order of import+target.

radical committed:
Use $(WasmBuildAppDependsOn) to ensure that we run the Prepare* task in the correct order

Commit:
Make sure the path is right for the MonoAOTCompiler in the runtime tests

Commit:
Merge remote-tracking branch 'steve/enable-wasm-aot-tests' into enable-wasm-aot-tests

naricc commented on src/tests/Common/wasm-test-runner/WasmTestRunner.proj:
```diff
@@ -10,6 +10,7 @@
     <NETCoreAppMaximumVersion>99.0</NETCoreAppMaximumVersion>
 
     <WasmAppBuilderTasksAssemblyPath>$(CORE_ROOT)\WasmAppBuilder\WasmAppBuilder.dll</WasmAppBuilderTasksAssemblyPath>
+    <MonoAOTCompilerTasksAssemblyPath>$(CORE_ROOT)\MonoAOTCompiler\MonoAOTCompiler.dll</MonoAOTCompilerTasksAssemblyPath>
```
So, we haven't actually made the changes necessary for the runtime tests to use AOT, right? Is this property used by the libraries tests?

steveisok replied:
In theory, if you set the right inputs, the runtime tests could take advantage.  I didn't set that up... Only the libraries tests.

naricc replied:
I guess my question was; does it belong in this PR? Because right now nothing is using it.

steveisok replied:
I think we add enabling it for the runtime tests later.

radical committed:
[wasm] console sample: track property name change

lewing commented on src/mono/netcore/sample/wasm/browser/Makefile:
```diff
@@ -7,12 +7,16 @@ else
 
 build:
-	$(DOTNET) build $(DOTNET_Q_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) WasmSample.csproj
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:Configuration=$(CONFIG) /p:TargetArchitecture=wasm /p:TargetOS=Browser $(MSBUILD_ARGS) WasmSample.csproj
```
Does this (realpath) work on OS/X without setting SHELL to a bash version newer than what it comes with?

akoeplinger replied:
we should only set EMSDK_PATH if it's not yet set (so I can use my own emsdk). If you do that as a Makefile variable you can easily resolve it to a full path with the `abspath` Makefile function:

```
EMSDK_PATH?=$(abspath $(TOP)/src/mono/wasm/emsdk)
```

marek-safar commented on src/mono/netcore/sample/wasm/browser/Program.cs:
```diff
@@ -2,12 +2,14 @@
 {
     public class Test
     {
+        [DynamicDependency(DynamicallyAccessedMemberTypes.All, "System.Runtime.InteropServices.JavaScript.Runtime", "System.Private.Runtime.InteropServices.JavaScript")]
```
Yep, this is wrong place.

Commit:
Remove DynamicDependency

akoeplinger commented on Directory.Build.targets:
```diff
@@ -21,4 +21,4 @@
     <NETCoreAppMaximumVersion>$(MajorVersion).$(MinorVersion)</NETCoreAppMaximumVersion>
   </PropertyGroup>
-</Project>
+</Project>
```
unrelated change.

akoeplinger commented on src/mono/netcore/sample/wasm/console/Makefile:
```diff
@@ -17,7 +17,7 @@ CONFIG?=Release
 
 build:
-	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) WasmSample.csproj
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) $(DOTNET_RUN_AOT_COMPILATION_ARGS) /p:TargetArchitecture=wasm /p:TargetOS=Browser /p:Configuration=$(CONFIG) $(MSBUILD_ARGS) WasmSample.csproj
```
same comment about EMSDK_PATH here.

akoeplinger commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -1,31 +1,84 @@
+      <Error Condition="'$(WasmMainAssemblyPath)' == ''" Text="%24(WasmMainAssemblyPath) property needs to be set" />
+      <Error Condition="'$(WasmAppDir)' == ''" Text="%24(WasmAppDir) property needs to be set" />
+      <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(WasmAOTDir)' == ''" Text="%24(WasmAOTDir) property needs to be set" />
+      <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
```
Hm I wonder if we still need to set this explicitly in the projects. We could resolve all the other paths that currently use it to use `%(RuntimePack.PackageDirectory)` instead. That's also basically what we'll need to do for "real" usage of this target outside of dotnet/runtime.

akoeplinger commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -38,9 +91,43 @@
 
+  <Target Name="_AfterWasmBuildApp" AfterTargets="_WasmBuildApp">
+    <!-- Run mono-cil-strip on the assemblies -->
+    <!-- TODO: replace this with a linker task -->
```
do you remember if we created an issue for that?

steveisok commented:
The remaining feedback will be addressed in a follow up PR
